### ISSUE TITLE
Implementation of authorization around the config-repo plugin extension

### ISSUE THREAD

**Issue by anishvenkat** [gocd/gocd#3636]
**Issue Type**

- Feature enhancement

##### Summary

Use case - Assuming that a  Go server has an environment `Production` consisting of some secure environment-variables which will be used as part of the pipelines that are part of this environment. A malicious user with commit rights to one of the configured config-repos could potentially create a pipeline mapped to `Production` environment which can have a task that could expose all the secure variables with a simple echo task.

Can we implement an authorization model around the config-repo plugin extension to allow/restrict certain actions and the entities upon which such an action could be performed?
@tomzo - any thoughts?

**Comment by tomzo** [comment]
The statement so far was that it should be handled by the plugin, rather then server side.
In the example above, it would be quite easy to either:
 * add a plugin setting (global for all repos) which contains some form of blacklist/whitelist of what is allowed.
 * add per-config-repo setting, such as

```xml
  <config-repos>
    <config-repo plugin="yaml.config.plugin-secure">
      <git url="repo1" />
      <configuration>
        <property>
          <key>forbidden_environment_references</key>
          <value>Production</value>
        </property>
      </configuration>
    </config-repo>
 </config-repos>
```
Then plugin would reject a config which assigns any pipeline to Production, when done in `repo1`. And I think it safe to assume that malicious user has only commit rights, but no access to XML.

I think it is much easier to add on plugin side than on server.

**Comment by mrmanc** [comment]
> I think it is much easier to add on plugin side than on server.

I think that is probably true but am I right in saying that the GoCD UI is currently not able to assign pipelines-as-code to environments? So if we forbade certain environments we would have no way to put these pipelines into those environments as admins.

I was thinking it would be nice to make sure that pipeline config and environment config were picked up from different places. I’d actually prefer that the pipeline config comes from the materials configured against the pipeline (appreciate there’s a chicken/egg thing going on there) rather than have to ask an admin to add a config repo via the UI. The environment configuration seems like a totally different use case and should be a different sort of repo so that a different bunch of people can manage it. That way we can continue to use code to describe what environment pipelines should reside in without creating a route around the current security around that. Something like this?

```
<config-repos>
    <pipeline-config-repo plugin="yaml.config.plugin-secure">
      <git url="repo1" />
    </pipeline-config-repo>
    <environment-config-repo plugin="yaml.config.plugin-secure">
      <git url="repo2" />
    </environment-config-repo>
 </config-repos>
```

**Comment by tomzo** [comment]
>  I right in saying that the GoCD UI is currently not able to assign pipelines-as-code to environments?

Yes. But it is not an UI limitation. You cannot make reference from XML element to *as-code* element. Other example is you cannot depend on pipeline-as-code from an UI-configured pipeline.
This limitation has good reasons, and I doubt that we will ever remove it.

> that pipeline config and environment config were picked up from different places [..] The environment configuration seems like a totally different use case and should be a different sort of repo so that a different bunch of people can manage it.

This can be also enforced using only plugin.
```xml
  <config-repos>
    <config-repo plugin="yaml.config.plugin-secure">
      <git url="repo1" />
      <configuration>
        <property>
          <key>only_pipelines</key>
          <value>true</value>
        </property>
      </configuration>
    </config-repo>
    <config-repo plugin="yaml.config.plugin-secure">
      <git url="repo2" />
      <configuration>
        <property>
          <key>only_environments</key>
          <value>true</value>
        </property>
      </configuration>
    </config-repo>
 </config-repos>
```

>  I’d actually prefer that the pipeline config comes from the materials configured against the pipeline 

You mean that you want to restrict pipelines as code being configured only by the same material which is used in pipeline materials?
I think that could be an option in plugin too.
There is already a [feature in JSON plugin](https://github.com/tomzo/gocd-json-config-plugin#configrepo), related to this. It can be added to yaml plugin too. Then plugin can limit acceptable pipelines to only those, which have `configrepo` as material.

**Comment by arvindsv** [comment]
A reason I was thinking of for this to be done on the server side, and not in the plugins is that it would be common to all plugins. Something like:

```
<config-repo plugin="yaml.config.plugin">
  <git url="repo1" />
  <actions>
    <allow>
<![CDATA[
environments:create
environment:edit:env1
pipeline:edit:pipeline1
pipelines:edit:pip-group1
]]>
    </allow>
  </configuration>
</config-repo>
```

That way, it would be enforced on the server side, during a merge and would not need to be done in each plugin. Of course, as you say, it is possible for the plugin to enforce it too.

**Comment by arvindsv** [comment]
And yes, one of the reasons to not allow a mix-and-match is that the config, by itself, would then be invalid, if the main XML could reference pipelines-as-code (which can go away).

**Comment by tomzo** [comment]
> A reason I was thinking of for this to be done on the server side, and not in the plugins is that it would be common to all plugins.

OK
 
```
<actions>
    <allow>
<![CDATA[
environments:create
environment:edit:env1
pipeline:edit:pipeline1
pipelines:edit:pip-group1
]]>
</allow>
```

I hope this is just a quick copy-paste from current auth model. It would have to be quite different for *as-code* entities. There are really no "actions" in *as code* approach, because it is declarative. E.g. what does `pipeline:edit:pipeline1` mean?
I think a query for allowed or not allowed entities would be more useful. With something like [json path](https://github.com/json-path/JsonPath).
After all, the plugin just returns a deep JSON object, which we can trim with some sort of filter
```json
{
    "pipelines" : [ ... ],
    "environments": [ ... ]
}
```
Then let's say I want to allow only pipelines:
```xml
<authorization>
    <allow>
<![CDATA[
$.pipelines.*
]]>
    </allow>
</authorization>
```

**Comment by arvindsv** [comment]
@tomzo - Yes, it was just meant to be a point for discussion. The actions I was thinking of were authorizations - as you said. The allowed and disallowed authorizations need to be things like:

- Allow/disallow a pipeline to be associated with environment "env1"
- Allow/disallow creating of a pipeline in group "pip-group-1"
- Allow/disallow a new environment to be created
- Allow/disallow a new pipeline group to be created
- Maybe allow/disallow specific materials? Maybe not needed.
- Allow/disallow setting up pipeline dependencies (and so, fetching artifact) from certain pipelines or groups

My point was, mainly, that adding flags like only_environments and only_pipelines might not be fine-grained enough.

I remember talking to you about something similar to this (not in as much detail), but can't seem to find it.

**Comment by kaushik-navi** [comment]
@arvindsv any thoughts / workarounds for this? I noticed you also mentioned this in https://github.com/gocd/gocd/issues/5175

**Comment by ckaushik** [comment]
@arvindsv any thoughts / workarounds for this? I noticed you also mentioned this in #5175

**Comment by arvindsv** [comment]
@ckaushik GoCD 19.11 released today adds support for policies. The policy support needs to be extended. Once that work is done, we could take a look at this. Initially, when I was thinking about it, I thought that the granular auth (policy) support can work for this. However, once I discussed it with people, it became clear that they are different. This needs different kinds of policies as mentioned above.

There isn't a workaround as far as I know. However, if you're up for making changes to the plugin, it should be fairly simple to accept a couple more configuration options and reject commits which make those changes. I'd played around with something like that, at one point: https://github.com/arvindsv/gocd-yaml-config-plugin/commit/63b331c87d1a872487438c27323c9a1a594b35fb

The right way to do it, as mentioned in this conversation is for GoCD to add support for it, so that it works across plugins.

**Comment by arvindsv** [comment]
FYI: There's a conversation relevant to this issue happening in #7605. Please take a look and feel free to make suggestions. Thanks!

**Comment by kritika-singh3** [comment]
This issue has been addressed with the introduction of rules around config-repos in GoCD version [20.2.0](https://www.gocd.org/releases/#20-2-0). Closing this issue.

Please feel free to open a new issue in case of a bug/concern.
